File: go/activity.go
package app

import (
	"context"
	"errors"
	"math/rand"
	"time"
)

type ExpensiveActivityParam struct {
	Input int
}

type ExpensiveActivityResult struct {
	Cost int
}

func ExpensiveActivity(ctx context.Context, param ExpensiveActivityParam) (ExpensiveActivityResult, error) {
	s := rand.Intn(param.Input)
	time.Sleep(50 * time.Millisecond * time.Duration(s))
	return ExpensiveActivityResult{Cost: s}, nil
}

type UnreliableActivityParam struct {
	Input int
}

func UnreliableActivity(ctx context.Context, param UnreliableActivityParam) error {
	s := rand.Intn(param.Input)
	if s < (3 * param.Input / 4) {
		return errors.New("unexpected error executing expensive activity")
	}
	return nil
}

File: py/activity.py
import asyncio
import random
from dataclasses import dataclass
from temporalio import activity


@dataclass
class ExpensiveActivityParam:
    input: int


@dataclass
class ExpensiveActivityResult:
    cost: int


@activity.defn
async def expensive_activity(param: ExpensiveActivityParam) -> ExpensiveActivityResult:
    s: int = random.randrange(param.input)
    await asyncio.sleep(50 / 1000 * s)
    return ExpensiveActivityResult(cost=s)


@dataclass
class UnreliableActivityParam:
    input: int


@activity.defn
async def unreliable_activity(param: UnreliableActivityParam):
    s: int = random.randrange(param.input)
    if s < (3 * param.input / 4):
        raise Exception("unexpected error executing expensive activity")
